src/Inciter/AMR/tet_store.h

```diff
@@ -0,0 +1,733 @@
+
+            void add(
+                    size_t id,
+                    tet_t nodes,
+                    Refinement_Case refinement_case,
+                    size_t parent_id
+            )
+            {
+
+                std::cout << "id " << id << " parent " << parent_id << std::endl;
+
+                add(id, nodes, refinement_case);
+
+                // Set parent id
+                master_elements.get(id).parent_id = parent_id;
+
+                master_elements.get(id).refinement_level =
+                    master_elements.get(parent_id).refinement_level+1;
+
+                // Deal with updating parent
+                master_elements.add_child(parent_id, id);
+
+                trace_out << "Added child " << id << " (" <<
+                    master_elements.get(parent_id).num_children << ")" << std::endl;
+            }
+
+            /**
+             * @brief Convenience function to add a tet to the master_elements
+             * and active_elements store
+             *
+             * @param nodes A list of the nodes which form th etet
+             * @param refinement_case The refinement case which caused this tet
+             * to be generated
+            */
+            void add(size_t id, tet_t nodes, Refinement_Case refinement_case)
+            {
+                store_tet(id, nodes);
+
+                size_t refinement_level = 0;
+                size_t parent_id = 0;
+
+                // Add to master list
+                master_elements.add(id, refinement_case, refinement_level, parent_id);
+
+                // The new master element should start as active
+                active_elements.add(id);
+            }
+
+            void add(tet_t nodes, Refinement_Case refinement_case)
+            {
+                size_t id = id_generator.get_next_tet_id();
+                add(id, nodes, refinement_case);
+            }
+
+            void add(
+                    size_t id,
+                    size_t first,
+                    size_t second,
+                    size_t third,
+                    size_t forth,
+                    Refinement_Case refinement_case,
+                    size_t parent_id
+                    )
+            {
+                return add(
+                        id,
+                        { {first, second, third, forth} },
+                        refinement_case,
+                        parent_id
+                        );
+
+            }
+
+            // NOTE: this does *not* deal with edges
+            /**
+             * @brief Function to delete a tet from the tet store (useful in
+             * derefinement)
+             *
+             * @param id id of the tet to delete
+             */
+            void erase(size_t id)
+            {
+                deactivate(id);
+                master_elements.erase(id);
+                tets.erase(id);
+                // TODO: Should this update the number of children here rather than at the call site?
```

Comment:
I think yes if the number of children is stored here, explicitly or inside a data structure that is stored here, i.e., in this class' state. No?